Specs/Overlay Support Spec.md

```diff
@@ -0,0 +1,120 @@
+
+## Solution Overview
+Overlays provide a flexible mechanism for modifying OpenAPI documents without directly modifying the original files. This allows necessary properties to be added, existing values modified, and the OpenAPI document tailored for a specific use case, for example developing a plugin manifest or adopting the document for a specific environment or a select audience group, while leaving the original openAPI document intact.
+
+## Functional Requirements
+1. **Overlay Support Feature**
+   - Support overlay files in accordance with the Overlay Specification (v1.0.0), available at [Overlay Specification](https://github.com/OAI/Overlay-Specification/blob/3f398c6/versions/1.0.0.md).
+   - Support both YAML and JSON formats for overlay files.
+   - Support addition, modification, and removal of properties to the target OpenAPI document.
+   - The target OpenAPI document should be provided separately in either JSON or YAML format. 
+   - Apply overlay file to an existing OpenAPI document to produce a hybrid of the overlay file and the original OpenAPI document without changing the original document.
+   - Ensure the resulting hybrid OpenAPI document remain valid according to the OpenAPI Specification.
+   - Overlay support feature will be added to the OpenAPI.NET library project as a new class. This will allow compatibility with current and future versions of OpenAPI.NET. This way developers enjoy the feature without adding their dependency footprint.
+   - Since OpenAPI.NET already supports both JSON and YAML, developers can use either format for their OpenAPI documents and overlays.
+   - The OverlayService class will validate the overlay file against OAI Overlay Specification and handle the modification of the OpenApiDocument by applying the overlay file. The modified objects will be used by the downstream logics as per the developer's needs.
+       
+  2. **User Journey**
+     1. Start by loading an existing OpenAPI document or creating one from scratch to create an instance of `OpenApiDocument`.
+        ```csharp
+        var openApiDocument = LoadOpenApiDocument(documentPath);
+
+     3. Use the OverlayService to apply the overlay file to the OpenAPI document. The OverlayService parses the overlay file, applies the changes directly to the OpenApiDocument instance, and performs in-place modifications.
+         ```csharp
+        var overlayService = new OverlayService();
+        overlayService.ApplyOverlay(openApiDocument, overlayPath);
```

Comment:
```suggestion
        var overlay = await Overlay.LoadAsync(stream);
        var jsonElement = await JsonDocument.ParseAsync(stream).RootElement;
        var newJsonElement = overlay.Apply(jsonElement);
```

Reply:
End to end process would be:
- Use OpenApiDocument.Load to load JSON/Yaml file with Overlay referenced in readerSettings
- Load Yaml and translate to JSONNodes or just load JSONNodes
- Load Overlay as JsonNodes.
- Apply Overlay to OpenAPI JsonNodes
- Pass "overlayed" JsonNodes to rest of OpenAPI Parser.